content/pytorch/concepts/nn/terms/loss-functions/loss-functions.md

```diff
@@ -0,0 +1,230 @@
+optimizer.zero_grad()
+loss.backward()
+optimizer.step()
+
+# Forward pass after optimization
+outputs = model(inputs)
+loss = criterion(outputs, targets)
+print('Loss after one optimization step:', loss.item())
+```
+
+Sample output for this example:
+
+```shell
+Initial Loss: 2.3014121055603027
+Loss after one optimization step: 2.294567823410034
+```
+
+## Example 3
+
+Binary Cross Entropy Loss (`nn.BCEWithLogitsLoss`) for Binary Classification:
+
+```py
+import torch
+import torch.nn as nn
+import torch.optim as optim
+
+# Hyperparameters
+input_size = 20
+hidden_size = 16
+learning_rate = 0.005
+batch_size = 16
+
+# Sample data
+inputs = torch.randn(batch_size, input_size)
+targets = torch.randint(0, 2, (batch_size, 1)).float()  # Binary targets (0 or 1)
+
+# Define a simple neural network
+model = nn.Sequential(
+    nn.Linear(input_size, hidden_size),
+    nn.ReLU(),
+    nn.Linear(hidden_size, 1)
+)
+
+# Define the loss function
+criterion = nn.BCEWithLogitsLoss()
+
+# Define the optimizer
+optimizer = optim.SGD(model.parameters(), lr=learning_rate)
+
+# Forward pass
+outputs = model(inputs)
+
+# Compute loss
+loss = criterion(outputs, targets)
+print('Initial Loss:', loss.item())
+
+# Backward pass and optimization
+optimizer.zero_grad()
+loss.backward()
+optimizer.step()
+
+# Forward pass after optimization
+outputs = model(inputs)
+loss = criterion(outputs, targets)
+print('Loss after one optimization step:', loss.item())
+```
+
+Sample output for the example:
+
+```shell
+Initial Loss: 0.6937122344970703
+Loss after one optimization step: 0.6932134628295898
+```
+
+> **Note**: When running these examples, the exact numerical values of the losses will vary each time due to random initialization of model weights and input data. The important aspect is observing the trend of the loss decreasing after the optimization step, indicating that the model is learning.
+
+## Common Loss Functions in `torch.nn`
+
+Some commonly used loss functions provided by PyTorch:
+
+- `nn.MSELoss`: Mean Squared Error Loss. Suitable for regression tasks.
+- `nn.CrossEntropyLoss`: Combines `nn.LogSoftmax` and `nn.NLLLoss`. Used for multi-class classification.
+- `nn.NLLLoss`: Negative Log Likelihood Loss. Often used with `nn.LogSoftmax`.
+- `nn.BCELoss`: Binary Cross Entropy Loss. Used for binary classification tasks.
+- `nn.BCEWithLogitsLoss`: Combines a **Sigmoid** layer and the BCELoss. More numerically stable.
+- `nn.L1Loss`: Mean Absolute Error Loss.
+- `nn.SmoothL1Loss`: A combination of `nn.L1Loss` and `nn.MSELoss`. Less sensitive to outliers.
+- `nn.HingeEmbeddingLoss`: Used for measuring similarity between samples.
+
+_Sigmoid function_ is a mathematical function that has an "S" shaped curve (sigmoidal shape). It is widely used in machine learning, particularly in logistic regression and neural networks, as an activation function to introduce non-linearity into the model and to map any real-valued number into a value between 0 and 1.
```

Comment:
We should skip explaining common loss functions here. This will make the entry longer, and beginners will not get an exact idea of what the function does.

These can be separately covered as term entries.


```suggestion
```